packages/utils/src/lib/server/cache/s3.ts

```diff
@@ -0,0 +1,171 @@
+import { BaseCache } from './base';
+import {
+	DeleteObjectCommand,
+	DeleteObjectsCommand,
+	GetObjectCommand,
+	HeadObjectCommand,
+	ListObjectsV2Command,
+	PutObjectCommand,
+	S3,
+	type S3ClientConfig,
+} from '@aws-sdk/client-s3';
+import { logger } from '../../logger.js';
+import { createJitter, JitterMode, type JitterFn } from '../../utils/misc.js';
+
+type S3CacheOptions = {
+	bucket: string;
+	keyPrefix?: string;
+	defaultTTL?: number;
+	defaultJitter?: JitterMode | JitterFn;
+};
+
+export class S3Cache<V extends Uint8Array = Uint8Array> extends BaseCache<V> {
+	readonly #options: S3CacheOptions;
+	readonly #client: S3;
+
+	public static init<V extends Uint8Array = Uint8Array>(
+		s3Options: S3ClientConfig,
+		options: S3CacheOptions,
+	): S3Cache<V> {
+		const client = new S3(s3Options);
+
+		return new this(options, client);
+	}
+
+	private constructor(options: S3CacheOptions, client: S3) {
+		super();
+
+		this.#options = options;
+		this.#client = client;
+	}
+
+	private buildKey(key: string): string {
+		return `${this.#options.keyPrefix ?? ''}${key}`;
+	}
+
+	public async get(key: string): Promise<V | undefined> {
+		const s3Key = this.buildKey(key);
+		try {
+			const head = await this.#client.send(
+				new HeadObjectCommand({
+					Bucket: this.#options.bucket,
+					Key: s3Key,
+				}),
+			);
+
+			const expiresAtStr = head.Metadata?.['expires-at'];
+			if (expiresAtStr && Date.now() > Number.parseInt(expiresAtStr, 10)) {
+				await this.delete(key);
+				return undefined;
+			}
+
+			const res = await this.#client.send(
+				new GetObjectCommand({
+					Bucket: this.#options.bucket,
+					Key: s3Key,
+				}),
+			);
```

Comment:
Why sending two requests (`HEAD` followed by `GET`) rather than just issuing a `GET`, checking the response metadata before reading the full body, and returning?

Obviously if the object obtained in the `GET` is expired the response should be closed without reading the body and the object should be deleted.